### PR TITLE
Incomplete memcard implementation

### PR DESCRIPTION
**Pull request by RedPanda4552**
WIP memcard implementation. Nothing operational but a certain someone convinced me to put it up here for visibility.

### CODE REVIEW

**Comment by Souzooka** [comment]
You'll want to use 4 spaces for indentation instead of tabs before this can be merged.

File: src/core/iop/memcard.cpp [hunk 1]
```diff
@@ -1,6 +1,308 @@
...
+
+	if (!file.is_open())
+	{
+		printf("[MCD] (WARN) Attempted to open ofstream on %s, but it cannot be opened! Aborting...", this->file_name);
+		return false;
+	}
+
+	memcpy(&raw_data[address], source.data(), source.size());
+	file.seekp(address);
+	file.write((char*) source.data(), source.size());
```
Comment (Souzooka) [inline src/core/iop/memcard.cpp]: Consider replacing the C-style cast with an explicit cpp cast, e.g. `static_cast<char*>(source.data())`

File: src/core/iop/memcard.cpp [hunk 2]
```diff
@@ -1,6 +1,308 @@
...
+	if (!is_erase_block_erased(16352 * bytes_per_page))
+	{
+		printf("[MCD] (WARN) Backup block 2 was NOT erased when this memcard was opened! The last write op on this memcard may have failed! The emulated PS2 should be restoring the old data soon, if it hasn't already...");
+	}
+
+	return true;
+}
+
+// Read length bytes from memcard starting at address into destination.
+bool Memcard::read(std::vector<uint8_t> destination, uint32_t address, uint32_t length)
```
Comment (Souzooka) [inline src/core/iop/memcard.cpp]: In this case, the function argument `destination` is copied when the function is called and as such program state is not altered. Should be `std::vector<uint8_t>& destination` instead.

File: src/core/iop/memcard.cpp [hunk 3]
```diff
@@ -1,6 +1,308 @@
...
+		printf("[MCD] (WARN) Attempted to read out-of-bounds memcard address %d! Aborting...", address);
+		return false;
+	}
+
+	memcpy(destination.data(), &raw_data[address], length);
+	return true;
+}
+
+// Write length bytes from source starting at address into memcard.
+bool Memcard::write(std::vector<uint8_t> source, uint32_t address)
```
Comment (Souzooka) [inline src/core/iop/memcard.cpp]: Recommend passing vector by reference, `std::vector<uint8_t>& source`

File: src/core/iop/memcard.cpp [hunk 4]
```diff
@@ -1,6 +1,308 @@
...
+	file.read(raw_data.data(), oem_memcard_bytes);
+
+	if (file)
+		printf("[MCD] (INFO) Successfully read all bytes into memory!");
+	else
+		printf("[MCD] (WARN) Failed to read all bytes into memory! Only read %d!", file.gcount());
+
+	file.close();
+	this->file_name = file_name;
+	this->formatted = true;
```
Comment (Souzooka) [inline src/core/iop/memcard.cpp]: This and the next few lines can be replaced with `this->formatted = std::equal(format_string.begin(), format_string.end(), raw_data);`

File: src/core/iop/memcard.cpp [hunk 5]
```diff
@@ -1,6 +1,308 @@
...
+	{
+		printf("[MCD] (WARN) Attempted to write out-of-bounds memcard address %d! Aborting...", address);
+		return false;
+	}
+
+	// Check if a bitwise and of source and address bytes would give a result
+	// different from the source byte. If so, that means something else
+	// currently exists here, or the erase block containing address has not been
+	// wiped yet. 
+	for (int i = 0; i < source.size(); i++)
```
Comment (Souzooka) [inline src/core/iop/memcard.cpp]: If I'm not mistaken, doesn't the function `is_erase_block_erased` serve this purpose?

**Comment by PSI-Rockin** [comment]
The author has stated he does not want to continue work on this PR as it has been too long since they touched it, so I'm closing it.

**Pull request closed without merging**